README.md
# SyntaxPane

[![Build Status](https://github.com/Sciss/SyntaxPane/workflows/Scala%20CI/badge.svg?branch=main)](https://github.com/Sciss/SyntaxPane/actions?query=workflow%3A%22Scala+CI%22)
[![Maven Central](https://maven-badges.herokuapp.com/maven-central/de.sciss/syntaxpane/badge.svg)](https://maven-badges.herokuapp.com/maven-central/de.sciss/syntaxpane)

## Statement

SyntaxPane is an extension to Java Swing's JEditorKit component which adds syntax highlighting support for various
languages, including Scala and Java. The original project JSyntaxPane can be found
[on google-code](http://code.google.com/p/jsyntaxpane/). This is a fork from the 0.9.6 branch.

The original project is (C)opyright by Ayman Al-Sairafi and released under the
[Apache License, Version 2.0](https://github.com/Sciss/SyntaxPane/raw/main/LICENSE).

All changes, reworkings and extensions in SyntaxPane (C)opyright 2011&ndash;2022 by Hanns Holger Rutz and
contributors. Released under that same license.

## Supported languages

Find here a list of all the
[supported languages.](https://github.com/Sciss/SyntaxPane/tree/main/src/main/jflex/de/sciss/syntaxpane/lexers) 

## Linking

The group-id and version have been adjusted to use my name space at Maven Central:

    "de.sciss" % "syntaxpane" % v

The current version `v` is `"1.2.1"`

## Building

JSyntaxPane builds with [sbt 1.x](http://www.scala-sbt.org/). The source code is purely Java, so no Scala compilation
is run. The project uses the [sbt-jflex](https://git.iem.at/sciss/sbt-jflex) plugin v0.4.0, which in turn uses
[JFlex](http://jflex.de/) 1.7.0 to generate the lexer Java sources for the supported languages.

To build run `sbt compile`. To run a demo application, run `sbt test:run`. Alternatively, you can build and run the
test with

    sbt package test:package
    java -cp target/syntaxpane-<version>.jar:target/syntaxpane-<version>-tests.jar de.sciss.syntaxpane.SyntaxTester

## Contributing

Please see the file [CONTRIBUTING.md](CONTRIBUTING.md) for details.

The people with the following GitHub handles and names have contributed (after the fork), thank you:
mattiaslundstrom, alexeyr, ShalokShalom, Qualtagh, karlvr, Patrick Guermonprez.


CHANGELOG.txt
Version 1.2.1
    * Fixes an issue HiDPI screens
    * Targets Java 1.8+ now

Version 1.2.0
    * Fixes an issue with OpenJDK 11 (@pguermo)
    * Builds with sbt 1.x now

Version 1.1.7
    * Add text/markdown

Version 1.1.6
    * Get default style for SyntaxView from config

Version 1.1.5
    * Fixes issues with Rhino loading in Java 8
    * Fixes issues with find and replace
    * Fixes issues with parentheses pairing
    * Improves focus when escaping dialogs

Version 1.1.4
    * Fixes issues with toggle-comment action

Version 1.1.3
    * Fixes various issues with line-numbers-ruler
    * Fixes some keyboard shortcuts on Mac and Linux
    * Fixes a NPE when not using scroll-pane
    * Fixes a bug with undo-buffer and copy/paste

Version 1.1.0
    * Rename to SyntaxPane
    * Use fully qualified package name de.sciss.syntaxpane
    * Enabling/disabling of Undo/Redo actions
    * Fix multiple OS X shortcuts
    * Fix some Scala related problems
    * All tracking of document's dirty state

Original JSyntaxPane Changelog:

Version 0.9.5 - Java Reflection Completions additions
    * Configuration class overhaul.  Each kit now has its own Configuration file
      which merges the superclass(es) Configurations.
    * Usable Reflect Completions Dialog (Java with F1 or Menu)
    * Java ENTER key responds properly to multi line comments
    * Font can be changed for each Kit
    * Kits for XHTML and Xpath added.  XHTML has a simple Preview Action.
    * Added Execute Script to JavaScript (can also be used for Groovy if Groovy
      Scripting support is installed properly).
    * Hotkeys for actions are displayed in Popup Menus
    * Word Completion action can also do CamelCase matches
    * Smart Home Action (and Smart Home Extend Selection)
    * Enhanced JavaScript support.  Issue 115
    * Better Undo grouping
    * Replace Dialog Enhancement.  Only for updatedable editors and added single replace.
    * Added append method to SyntaxDocument
    * Fixed Issue 130 - Compound Undo on multiple lines.
Version 0.9.4 - IntelliSense additions
    * Adding IntelliSense to Java with simple List of selectable keywords
    * Added Toggle Comment Actions using Control SLASH
    * Fixed Issue 47.
    * Added Clojure, Scala, DOS Batch and 'nix bash support
    * Added Configurable Popup menus with default Tango Desktop icons
    * Added configurable format for the CaretMonitor class
    * Toggle Comments Action selects the lines affected after being performed
    * Added multi-line support in ActionUtils.insertMagicString method.
    * GotoLine dialog responds to ESC key
Version 0.9.3 - Start of scripts for the document:
    * Added new methods getLine() to SyntaxDocument
    * SyntaxDocument getIndexOf methods deprecated, use getMatchers instead
    * Added Line Numbering to Java
    * Added CaretMonitor Class
    * Merged Find And Replace dialogs into one.
    * Created SyntaxComponent interface and have all UI components implement
      that interface.  DefaultSyntaxKit will use config.properties class to
      dynamically install these components.
    * Added Right Margin option and Single Color Selection Options
    * Added Python, C and C++ Support
    * Added Ruby Syntax Support
    * Fixed Issue 37 (NPE for LineNumbersRuler)
    * Fixed Issue 39 (Highlighting Tokens overrides selection highlights)
    * Fixed some JavaDoc comments.
    * Fixed Line Numbers being displayed for the height of the editor and now
      just for the actual available lines.
    * Fixed Margin typo in all project. Issue 43
    * Changing Actions to be more configurable:
      * SyntaxActions renamed to ActionUtils
      * Removed all inner classes from SyntaxActions
      * Will create new SyntaxAction interface that will allow dynamic addition
        actions (in the addKeyActions of DefaultSyntaxKit
    * Added Text AA property to SyntaxView
Version 0.9.2:
    * Fixing Java Indentation and Un-Indentation Actions
    * Added and used (in the Tester) clearUndoes on the SyntaxDocument
    * Added Basic JFlex Syntax
    * Added getContentTypes to DefaultSyntaxKit to get all registered
      ContentTypes.  This is now also used in the SyntaxTester instead of
      hardcoding the types.
    * Cleanup and optimization of Lexers by removing duplicate Java Code
      (replaced with Regex OR)
    * Removed calls to deprecated calls in SyntaxView
    * Modified the SyntaxDOcument to override the fireXXX methods and parse
      the document at that time instead of overriding the inserUpdate method.
      This fixes issue 24.
    * Added Groovy GString expression highlights
    * Removed getLanguages method from Lexer interface and implementations.
    * Added WARNING and ERROR TokenTypes and added their default styles.
    * Moved Keymaps and install methods to SyntaxKits instead of the Lexers
    * Removed deprecated methods from SyntaxActions class
    * Split SyntaxActions into new package and moved inner classes to the new
      package
    * Added Token HighLighter to Java
    * Added Pairs Highlighter to Java
    * Token class made immutable (all final fields)
    * Added Find and Replace Dialogs and Action to Java (mapped to C-F and
      C-H )
    * Added pair matching to XML tags
    * Added CDATA matching and pair highlights for XML
    * Added Comment Pair Highlighting in XML
    * Fixed highlighting with selections so the selection always appears
    * Merged Find and Replace Dialogs into one
    * Fixed issue 33 (undoable to add new line after final closing brace for
      Java)
    * Added Groovy Multi-Line strings and fixed Comments as Regex issue.
    * Fixes to empty find text field causing NPE
Version 0.9.1
    * Fixed empty strings in XML syntax  Issue 29
    * Fixed TAL lexer using incorrect package name
Version 0.9.0
    * Initial version after major overhaul


src/main/java/de/sciss/syntaxpane/TokenConstants.java
/*
 * Copyright 2008 Ayman Al-Sairafi ayman.alsairafi@gmail.com
 * Copyright 2011-2022 Hanns Holger Rutz.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License
 *       at http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package de.sciss.syntaxpane;

/**
 * Constants used by Tokens.
 * @author Ayman Al-Sairafi, Hanns Holger Rutz
 */
public class TokenConstants {
    /**
     * Token Kinds
     */
    // Operators and separators - Suitable for Java Type Languages
    public static final short EQ            =  1;   // =
    public static final short GT            =  2;   // >
    public static final short LT            =  3;   // <
    public static final short NOT           =  4;   // !
    public static final short COMP          =  5;   // ~
    public static final short QUESTION      =  6;   // ?
    public static final short COLON         =  7;   // :
    public static final short EQEQ          =  8;   // ==
    public static final short LTEQ          =  9;   // <=
    public static final short GTEQ          = 10;   // >=
    public static final short NOTEQ         = 11;   // !=
    public static final short ANDAND        = 12;   // &&
    public static final short OROR          = 13;   // ||
    public static final short PLUSPLUS      = 14;   // ++
    public static final short MINUSMINUS    = 15;   // --
    public static final short PLUS          = 16;   // +
    public static final short MINUS         = 17;   // -
    public static final short MULT          = 18;   // *
    public static final short DIV           = 19;   // /
    public static final short AND           = 20;   // &
    public static final short OR            = 21;   // |
    public static final short XOR           = 22;   // ^
    public static final short MOD           = 23;   // %
    public static final short LSHIFT        = 24;   // <<
    public static final short RSHIFT        = 25;   // >>
    public static final short URSHIFT       = 26;   // >>>
    public static final short PLUSEQ        = 27;   // +=
    public static final short MINUSEQ       = 28;   // -=
    public static final short MULTEQ        = 29;   // *=
    public static final short DIVEQ         = 30;   // /=
    public static final short ANDEQ         = 31;   // &=
    public static final short OREQ          = 32;   // |=
    public static final short XOREQ         = 33;   // ^=
    public static final short MODEQ         = 34;   // %=
    public static final short LSHIFTEQ      = 35;   // <<=
    public static final short RSHIFTEQ      = 36;   // >>=
    public static final short URSHIFTEQ     = 37;   // >>>=
    public static final short LPAREN        = 38;   // (
    public static final short RPAREN        = 39;   // )
    public static final short LBRACE        = 40;   // {
    public static final short RBRACE        = 41;   // }
    public static final short LBRACK        = 42;   // [
    public static final short RBRACK        = 43;   // ]
    public static final short SEMICOLON     = 44;   // ;
    public static final short COMMA         = 46;   // ,
    public static final short DOT           = 47;   // .

    // Keywords for Java Type Languages
    public static final short KW_START          = 255;
    public static final short KW_abstract       = KW_START +  0;
    public static final short KW_assert         = KW_START +  1;
    public static final short KW_boolean        = KW_START +  2;
    public static final short KW_break          = KW_START +  3;
    public static final short KW_byte           = KW_START +  4;
    public static final short KW_case           = KW_START +  5;
    public static final short KW_catch          = KW_START +  6;
    public static final short KW_char           = KW_START +  7;
    public static final short KW_class          = KW_START +  8;
    public static final short KW_const          = KW_START +  9;
    public static final short KW_continue       = KW_START + 10;
    public static final short KW_do             = KW_START + 11;
    public static final short KW_double         = KW_START + 12;
    public static final short KW_else           = KW_START + 13;
    public static final short KW_extends        = KW_START + 14;
    public static final short KW_final          = KW_START + 15;
    public static final short KW_finally        = KW_START + 16;
    public static final short KW_float          = KW_START + 17;
    public static final short KW_for            = KW_START + 18;
    public static final short KW_default        = KW_START + 19;
    public static final short KW_implements     = KW_START + 20;
    public static final short KW_import         = KW_START + 21;
    public static final short KW_instanceof     = KW_START + 22;
    public static final short KW_int            = KW_START + 23;
    public static final short KW_interface      = KW_START + 24;
    public static final short KW_long           = KW_START + 25;
    public static final short KW_native         = KW_START + 26;
    public static final short KW_new            = KW_START + 27;
    public static final short KW_goto           = KW_START + 28;
    public static final short KW_if             = KW_START + 29;
    public static final short KW_public         = KW_START + 30;
    public static final short KW_short          = KW_START + 31;
    public static final short KW_super          = KW_START + 32;
    public static final short KW_switch         = KW_START + 33;
    public static final short KW_synchronized   = KW_START + 34;
    public static final short KW_package        = KW_START + 35;
    public static final short KW_private        = KW_START + 36;
    public static final short KW_protected      = KW_START + 37;
    public static final short KW_transient      = KW_START + 38;
    public static final short KW_return         = KW_START + 39;
    public static final short KW_void           = KW_START + 40;
    public static final short KW_static         = KW_START + 41;
    public static final short KW_while          = KW_START + 42;
    public static final short KW_this           = KW_START + 43;
    public static final short KW_throw          = KW_START + 44;
    public static final short KW_throws         = KW_START + 45;
    public static final short KW_try            = KW_START + 46;
    public static final short KW_volatile       = KW_START + 47;
    public static final short KW_strictfp       = KW_START + 48;
}
